### DISCUSSION TITLE
AttributeError: 'module' object has no attribute 'dupterm'

### DISCUSSION THREAD

**Discussion by cve2022** [micropython/micropython#10662] [RP2040 / Pico] [answered]
I was going to experiment with dupterm to try to understand how it works and exactly what it is for and how I could use it (or try posting not dumb questions in the forum AFTER I did not sort it out :smile:), but I got the error above when trying to instantiate it.

From an REPL in rshell:
```
MicroPython v1.19.1-782-g699477d12 on 2022-12-24; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>>
>>> import os
>>> dir(os)
['__class__', '__name__', 'remove', '__dict__', 'VfsFat', 'VfsLfs2', 'chdir', 'getcwd', 'ilistdir', 'listdir', 'mkdir', 'mount', 'rename', 'rmdir', 'stat', 'statvfs', 'umount', 'uname', 'unlink', 'urandom']
>>>
```
FWIW, I compiled it from source code.

OK, just in case I had done something wrong when I compiled it, I just installed the latest nightly build and... wait for it...
```
D:\Users\elder\Desktop\micropython\Testes>rshell -p COM8 --buffer-size 512 repl
Using buffer-size of 512
Connecting to COM8 (buffer-size 512)...
Trying to connect to REPL  connected
Retrieving sysname ... rp2
Testing if ubinascii.unhexlify exists ... Y
Retrieving root directories ... /blink.py/ /blink2.py/ /main.py/ /neopix_tst.py/ /tst_poll1.py/ /tst_poll3.py/ /tst_uart.py/ /tst_uart2.py/ /x.py/ /y.py/
Setting time ... Feb 04, 2023 12:15:15
Evaluating board_name ... pyboard
Retrieving time epoch ... Jan 01, 1970
Entering REPL. Use Control-X to exit.
>
MicroPython v1.19.1-852-g9ea64a36a on 2023-02-03; Raspberry Pi Pico with RP2040
Type "help()" for more information.
>>>
>>> import os
>>> dir(os)
['__class__', '__name__', 'remove', '__dict__', 'VfsFat', 'VfsLfs2', 'chdir', 'getcwd', 'ilistdir', 'listdir', 'mkdir', 'mount', 'rename', 'rmdir', 'stat', 'statvfs', 'umount', 'uname', 'unlink', 'urandom']
>>>
```

So... I is not compiled by default, not supported at all on the Pico or what?

And, as I am on it, why is it exactly for?

**Comment by robert-hh** [comment, marked as answer]
dupterm in only included for network capable boards. If you modify mpconfigport.h around line 155 and move the DUPTERM setting out of the `#if MICROPY_PY_NETWORK` block, then you'll have dupterm and can e.g. mirror REPL to a UART.

**Reply by cve2022** [reply]
Thanks, Robert. IOW, I do not need it for serial ports/stdin/stdout. For some reason, I thought it had something to do with these interfaces.

**Reply by freemansoft** [reply]
`dupterm` isn't available on the base Pico without networking?  That is an interesting constraint. I use dupterm for Bluetooth CDC for boards without a network connection..  

Will a PR be approved if we move  `MICROPY_PY_OS_DUPTERM ` outside the if block?

It looks like only some ports suffer from this.
* The default file - i couldn't find with the `MICROPY_PY_NETWORK`  variable
* The RP2 mpconfigport.h has this network switch https://github.com/micropython/micropython/blob/9e9176467179bcb477b92f32033e62fcde14ea41/ports/rp2/mpconfigport.h
* The STM32 alwayse enables it https://github.com/micropython/micropython/blob/9e9176467179bcb477b92f32033e62fcde14ea41/ports/stm32/mpconfigport.h
* The SAMD file mpconfigport.h always enables it https://github.com/micropython/micropython/blob/17ab2f671b1abe93ccd0024c6750a63e0c20ee6a/ports/samd/mpconfigport.h
* The ESP8266 file mpconfigport.h enables it, probably because all ESP8266 have networking https://github.com/micropython/micropython/blob/8f4c108025d4590a1b7f9d062ad64f95c45e0bc2/ports/esp8266/mpconfigport.h
* The ESP32 file mpconfigport.h always enables it, do all ESP32 have networking? https://github.com/micropython/micropython/blob/c7301b8d03d4681247c892d260696b0bfa4139d9/ports/esp32/mpconfigport.h

```
MICROPY_PY_OS_DUPTERM
```

**Comment by robert-hh** [comment]
Usually, stdin/stdout goes to USB, unless you configured your build for serial. If you want to duplicate the USB stdin/stdout to UART, then you need dupterm.

**Reply by cve2022** [reply]
OK, so I did not entirely misanderstood it.

In the project I am working where I am using the Pico and MP, I will need interaction with a terminal emulator (e.g. putty) to show a menu. I want do do it through the USB virtual serial port, through stdin/stdout. I have made some experiments with a tiny script (code below). 

```
import sys
import select
from machine import Pin
import utime as time

led=Pin(25, Pin.OUT)
led.on()
spoll=select.poll()        # Set up an input polling object.
spoll.register(sys.stdin,select.POLLIN)    # Register polling object.

# Function to read a character from USB serial or return None.
def kbhit():
    return sys.stdin.buffer.read(1) if spoll.poll(0) else None

while True:
    x = kbhit()
    led.toggle()
    if x:
        sys.stdout.write(hex(ord(x))+' ')
```
On the PCB I designed for the project, I have added headers to both UARTs of RP2, but I would hardly need them both. I intend leaving UART0 alone to have the benefit of debugging and other development activities as well as for powering the board. I may use UART1 for the aforementioned menu as a plan B (which I hope I do not need) but then I could use it through normal UART access and not necessarily through stdin/stdout (though it is nice knowing I could).
